src/backend/joanie/core/models/products.py

```diff
@@ -376,11 +379,72 @@ def can_be_state_validated(self):
         An order can be validated if the product is free or if it
         has invoices.
         """
-        return self.total == 0 or self.invoices.count() > 0
+        return self.total == 0.0 or self.invoices.count() > 0
+
+    def can_be_state_submitted(self):
+        """
+        An order can be submitted if the order has a course, an organization,
+        an owner, and a product
+        """
+        return (
+            self.course is not None
+            and self.organization is not None
+            and self.owner is not None
+            and self.product is not None
+        )
+
+    @transition(
+        field="state",
+        source=enums.ORDER_STATE_DRAFT,
+        target=enums.ORDER_STATE_SUBMITTED,
+        conditions=[can_be_state_submitted],
+    )
+    def _submit(self, billing_address=None, credit_card_id=None, request=None):
+        """
+        Transition order to submitted state.
+        """
+
+        if self.total == 0.0:
+            return None
+        if billing_address is None:
+            raise ParseError({"billing_address": ["This field is required."]})
+        payment_backend = get_payment_backend()
+        if credit_card_id:
+            try:
+                credit_card = payment_models.CreditCard.objects.get(
+                    owner=self.owner, id=credit_card_id
+                )
+                payment_info = payment_backend.create_one_click_payment(
+                    request=request,
+                    order=self,
+                    billing_address=billing_address,
+                    credit_card_token=credit_card.token,
+                )
+            except (payment_models.CreditCard.DoesNotExist, NotImplementedError):
+                pass
+        else:
+            payment_info = payment_backend.create_payment(
+                request=request, order=self, billing_address=billing_address
+            )
+
+        return payment_info
+
+    def submit(self, billing_address=None, credit_card_id=None, request=None):
+        """
+        Transition order to submitted state and to validate if order is free
+        """
+        payment_info = self._submit(billing_address, credit_card_id, request)
+        if self.total == 0.0:
+            self.validate()
+        return payment_info
```

Comment:
if `self.total == 0.0` I don't think we should call `self._submit` mainly to ease readability.